pandas/core/groupby/groupby.py

```diff
@@ -795,13 +795,13 @@ def _try_cast(self, result, obj, numeric_only: bool = False):
                 # Ensure we localize to UTC first before converting
                 # to the target timezone
                 arr = extract_array(obj)
-                try:
+                if not is_datetime64tz_dtype(result.dtype):
+                    # arr may have already been cast back to tzaware in _try_cast
                     result = arr._from_sequence(result, dtype="datetime64[ns, UTC]")
                     result = result.astype(dtype)
-                except TypeError:
```

Comment:
I *suppose* you can end up here with a different dtype that is not convertible to datetime64 (there is no guarantee that your groupby function or aggregation returns the same dtype as the original column). In such a case, we should not raise the error, but rather fall back to the original `result` as done now?

Reply:
AFAICT the relevant functions are all dtype-preserving (well, e.g. sometimes int can be cast to float, but not relevant for datetime64).  Is there a case that im forgetting?

Reply:
What do you mean with "the relevant functions" ? I think this gets used for user defined functions as well?

Reply:
most of the calls that get here go through our cython functions, but you're right that we can also get here from _python_agg_general.  But in the case of user defined functions that may not be dtype-preserving, we shouldn't be doing this casting at all, regardless of whether we pre-check vs catch